New Module: digital_ocean_domain_record - manage domain records

woohgit opened the pull request:
##### SUMMARY
This module (digital_ocean_domain_record) allows the use of the DigitalOcean REST API (https://developers.digitalocean.com/documentation/v2/) to create and delete DigitalOcean domain records.

Originally I added the update feature also, but during the integration testing phase I felt like it's not a good usecase.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
digital_ocean_domain_record

##### ANSIBLE VERSION
```
ansible 2.5.0 (cloud/digitalocean/domain_record 6ed546fddb) last updated 2017/10/16 08:53:24 (GMT +900)
  config file = None
  configured module search path = [u'/home/wooh/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/wooh/repos/ansible/lib/ansible
  executable location = /home/wooh/repos/ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION
Management of the DigitalOcean Domain Records is a long-time missing feature. Although I don't like depending on external libraries like `dopy` or `python-digitalocean`. 

With the re-usage of the `DigitalOceanHelper` makes it much more easier.

woohgit committed:
New Module: digital_ocean_domain_record - manage domain records

BondAnthony commented:
Hi @woohgit! We appreciate the contribution and the interest in removing `dopy` as a module dependency. Currently the `digital_ocean_domain` module is being refactored to not use `dopy` in PR #29179. We also have PR #30201 to include record management functionality in the `digital_ocean_domain` module. Do you have any issues with including this functionality within the `digital_ocean_domain` module once the `dopy` dependency is removed? Thank you again for the contribution!

woohgit commented:
@BondAnthony Thanks for your quick reply.

I'm not sure if we really want to "merge" the `domain` and the `domain_record` functionality. From my experience I think keeping them as a separate `plugin` make them easier to read / write.

But if eveyone votes on including this in the `digital_ocean_domain` I'm happy to do that. Originally my plugin created the domain if it was missing, but I removed that functionality.

I'm happy to review PR #29179 and contribute to it if no one is against it. The question is: how common is that a plugin can return multiple data types like the `domain` object or the `domain_record` object. Is it supported/common in ansible and if yes, how easy to document it? 

Maybe adding an extra parameter to the `digital_ocean_domain` plugin as `records`? But how one would `delete` a record if we have only one `state` parameter? Maybe having another `state` parameter inside of the `records` parameter would do the trick like:

```yaml
- name: Create domain with records
  digital_ocean_domain:
    name: example.com
    ip: 127.0.0.1
    state: present
    records:
      - www:
          name: www
          data: "@"
          type: CNAME
          ttl: 3600
          state: absent
      - mail:
          name: mail
          data: mail.example.com
          type: MX
          state: present
````

wdyt? I will send a PR to your `BondAnthony:do_refactor/testing` branch.

woohgit committed:
pep8 and pylint changes

woohgit committed:
cmp workaround for python3

ryansb commented on lib/ansible/modules/cloud/digital_ocean/digital_ocean_domain_record.py:
```diff
@@ -0,0 +1,360 @@
+  type:
+    description:
+     - The type of record you would like to create.
+    choices: [ A, AAA, CNAME, NS, TXT, MX, SRV ]
```
I think it's `A` and `AAAA`, not `AAA`

woohgit replied:
Correct, the documentation part has this typo.

ryansb reviewed (changes requested):
From a usage perspective, it seems like if I had, say, a site with `www.my.site A 123.123.123.123`, then my service IP changes to `www.my.site A 234.234.234.234`, this module would create a *new* A record and leave the existing one. Is that right?

ryansb commented:
@woohgit I'd be :-1: on merging this function into the existing domain module, since usually people won't have one single task that knows *all* the records that must exist.

BondAnthony commented:
@ryansb your suggesting to make a new module for managing domain records only and keeping the digital_ocean_domain module as is... Minus the `dopy` dependency.

woohgit commented:
#### Update record

@ryansb While creating this module I implemented the `update_record` feature too. I had the `force_update` parameter to update the records if the `name`, `type` matches. But since we have a lot of record types, it's absolutely not intuitive for the user which one gets updated, which one will be created and which one will be ignored (`changed=False`)

I finally checked all other DNS manager plugins/tools and finally decided not ot include the `update_record` feature because just like the `file' module or the `package' module no one will remove the package or remove the file if I change the file path or change the package name. 

Although adding back the update record feature with a `force_update` parameter make sense and I'm absoultely not against it. There is a specific use case when you have multiple `name` and `type` matches and can`t decide which one to update, but it's easy: only update only if there is only 1 exactly matching record.

#### Including vs not including in DigitalOcean Domain plugin

I personally don't like the idea of adding it to the `digital_ocean_domain` plugin, so I'm ok with keeping the `digital_ocean_domain_records` module as a separate plugin.

woohgit committed:
Fix typo in doc yaml

woohgit committed:
Change another AAA typo

BondAnthony commented on lib/ansible/modules/cloud/digital_ocean/digital_ocean_domain_record.py:
```diff
@@ -0,0 +1,360 @@
+    domain: example.com
+    type: A
+    name: www
+    ip: 127.0.0.1
```
`ip` should be data, right?

woohgit replied:
Correct, fixed.

woohgit committed:
Fix documentation typo

BondAnthony commented:
@woohgit I appreciate the conversation and the push to keep this separate from the digital_ocean_domain module, both of you have valid points! I've tested this module out and it works as designed. Thank you for development and please feel free to contribute new modules or assist with other already open PR.

woohgit commented:
@BondAnthony Thanks and I'm happy to hear that. 

This weekend I'll have time to review the code and add the update functionality to it.

BondAnthony commented:
Sounds great!

Thank you @ryansb! abadger actually sent your irc response to me from the other day. Thanks!

woohgit commented:
@ryansb @BondAnthony 

What should an `update_record` return if there are multiple records with the same `type` and `name` like multiple `MX` records or multiple `A` records with the same `www` name but with a different IP? DigitalOcean allows this, so it works like a `round-robin` IIRC.

Shall we return a `module.exit_json(changed=False, result=result)` or rather a `module.fail_json(msg="xxx")`? 

I think the `changed=False` would be a natural response. But maybe letting him know that updating the record is not possible (because it's not possible to determine which record to update) is also important.

BondAnthony commented:
I would agree with the `changed=False` since you won't know which record to update, along with showing them the results. However, you should probably fail the tasks because your in a state that can't be handled without more information. Gracefully failing but allowing the user to continue on could lead to confusion. We should still provide a way to update the records. I believe the only way to accomplish this update would be to have the user provide the record id. 

I was curious to know how DO handles this case with their own CLI. They actually force the users to provide the record id to make these updates.

```
$> doctl compute domain records update butter.toilops.com --record-name ww --record-data 192.168.200.31
Error: id is invalid because cannot be less than 1

$> doctl compute domain records update butter.toilops.com --record-name www  --record-id 29396280 --record-data 192.168.200.31
```

Output from the ip when you have duplicate type/name.
```
        {
            "id": 29396195,
            "type": "A",
            "name": "www",
            "data": "192.168.200.2",
            "priority": null,
            "port": null,
            "ttl": 3600,
            "weight": null,
            "flags": null,
            "tag": null
        },
        {
            "id": 29396280,
            "type": "A",
            "name": "www",
            "data": "198.168.200.3",
            "priority": null,
            "port": null,
            "ttl": 3600,
            "weight": null,
            "flags": null,
            "tag": null
        }
```

woohgit commented:
@BondAnthony wrt `requirements` I have a question.

In ansible we can say, the `type/name/data` is required if `state=present`. This is pretty nice.

```python
required_if=[('state', 'present', ('type', 'name', 'data'))]
```

But can I define the same for `state=absent` but with a tiny twist. Require `record_id` or `name/type/data`? I know that there is a `require_one_of` so basically what I'd need is:

```python
required_one_of=['record_id', ('type', 'name', 'data')]
```
I'd love to define the requirements in a neat / beautiful way make sure to handle the missing parameters withouth coding the logic.

woohgit committed:
Refactoring based on PR conversation

- add feature: update record if `force_update=yes` is set.
- add feature: delete/update based on `record_id`
- fail if can't decide which one to update (no record_id is given)
- reduced calls
- encapsulate it to `DigitalOceanRecordDomainManager`
- change method visibility to private if it's private

woohgit committed:
Improve/fix documentation

woohgit commented:
I refactored it a little bit and did some testing. Mostly this yaml covered all the things I wanted to test:

```yaml
- name: Create default A
  digital_ocean_domain_record:
    oauth_token: "{{ oauth_token }}"
    domain: "{{ domain }}"
    data: "{{ otaku_ip }}"
    type: A
    name: "@"

- name: Create google MX records
  digital_ocean_domain_record:
    oauth_token: "{{ oauth_token }}"
    domain: "{{ domain }}"
    priority: "{{ item.priority }}"
    data: "{{ item.data }}"
    type: MX
  register: mx
  loop:
    - {priority: 10, data: "aspmx.l.google.com"}
    - {priority: 20, data: "alt1.aspmx.l.google.com"}
    - {priority: 20, data: "alt2.aspmx.l.google.com"}

- name: Create www
  digital_ocean_domain_record:
    oauth_token: "{{ oauth_token }}"
    domain: "{{ domain }}"
    name: www
    type: A
    data: 127.0.0.1

- name: Adjust www based on name/type
  digital_ocean_domain_record:
    oauth_token: "{{ oauth_token }}"
    domain: "{{ domain }}"
    name: www
    type: A
    data: 127.0.0.2
    force_update: yes

- name: Add another www with a different IP
  digital_ocean_domain_record:
    oauth_token: "{{ oauth_token }}"
    domain: "{{ domain }}"
    name: www
    type: A
    data: 127.0.0.3

- name: Update MX record based on record_id
  digital_ocean_domain_record:
    record_id: 29509199
    oauth_token: "{{ oauth_token }}"
    domain: "{{ domain }}"
    priority: "{{ item.priority }}"
    data: "{{ item.data }}"
    type: MX
    force_update: yes
  loop:
    - {priority: 15, data: "aspmx.l.google.com"}

- name: Delete record based on name/type/data
  digital_ocean_domain_record:
    oauth_token: "{{ oauth_token }}"
    domain: "{{ domain }}"
    priority: "{{ item.priority }}"
    data: "{{ item.data }}"
    type: MX
    state: absent
  loop:
    - {priority: 20, data: "alt2.aspmx.l.google.com"}

- name: Delete record based on record_id
  digital_ocean_domain_record:
    oauth_token: "{{ oauth_token }}"
    domain: "{{ domain }}"
    state: absent
    record_id: 29509199

- name: Can't update because more than 1 similar
  digital_ocean_domain_record:
    oauth_token: "{{ oauth_token }}"
    domain: "{{ domain }}"
    priority: "{{ item.priority }}"
    data: "{{ item.data }}"
    type: MX
    force_update: yes
  loop:
    - {priority: 17, data: "aspmx.l.google.com"}
```

Only the last one should fail because there are multiple records with the same type/name. (of course the record_id should be valid to update/delete)

BondAnthony commented:
@woohgit you can use required_if for both state=absent and state=present.

```
    required_if=[
        ('state', 'present', ('type', 'name', 'data')),
        ('state', 'absent', ('record_id', 'data'))
    ]
```

Do you have any other thoughts on how to resolve updates when records are similar?

woohgit commented:
@BondAnthony 

I think the best would be:

- if DO allows to create multiple name/type with different data, lets try to create it, except if the `force_update`. AFAIR it allows multiple A/AAAA/MX records, but not CNAME records for example.
- If `force_update` is set and there are multiple name/type matching, give error.
- If `force_update` is set, and there is only 1 matching, update that record.
- If `force_update` is set and `record_id` is also set, update the specific record.

So the plan would be to make sure (via doctl for example) what `type/name` pairs are allowed to be created multiple times.

woohgit commented:
Uh long time haven't had time to check this. I'll have a look this weekend and verify / sync the things we agreed on. Stay tuned..

woohgit commented:
@BondAnthony I slept on it and came to a conclusion. What do you think about it?

Maybe we don't really want to have the `update` feature. We only want to have the `present` and `absent`. We might want to enforce that all records with `present` are created and everything else should be *deleted*. So if you remove a domain record without specifying an `absent` action on the record, it should be deleted by this module.

The module should ensure that all domain records are `created` and exist. If a domain record has changed, it could be just deleted and re-created OR just simply update if it's easy and possible, but I see this operation rather a delete/recreate. I don't really like now the idea of the `update` or `force_update`. They don't really add any value just confuses everyone.

Any comments?

woohgit commented:
@jeromecc The problem is that you can add multiple records with the same subdomain. And it is hard to determine WHEN to update the record and when to create a new one. Because sometimes you want to create a round-robin record with muliple entries (www -> 1.1.1.1, 1.1.1.2,1.1.1.3), but sometimes you only want to update your records (www: 1.1.1.1 -> www -> 1.1.1.2). We need an extra flag for determining if this is an update request or a new request. 

Actually it's not HARD to determine, but all have to agree on what the software supposed to do at specific states. 

Additionally DO allows to create the exact same records for some types but not for others. And we have to be consistent with that too.

You're new to GitHub right? This is a `branch` and it is in my https://github.com/woohgit/ansible/tree/cloud/digitalocean/domain_record repository on github. The branch name is `cloud/digitalocean/domain_record`

jeromecc commented:
Thank you for your quick answer.

Additional weird stuff about DO domain name and hostname record management. Say you own example.com domain name. You register it with DO and you get example.com in your domains list.
You can create an A record named "foo" with IP 1.2.3.4
Then you use their interface (or digital_ocean_domain  which is wrongly described as "Create/delete a DNS record in DigitalOcean" in Ansible documentation here https://docs.ansible.com/ansible/2.5/modules/digital_ocean_domain_module.html) to create another domain called... foo.example.com and create an  "@ A record" with ip 4.5.6.7
Then I don't know exactly what happens about resolution. But this situation happened to me because of digital_ocean_domain Ansible module. I will open an issue about this documentation bug.

I will try to think more about this problem and contribute.

phemmer commented on lib/ansible/modules/cloud/digital_ocean/digital_ocean_domain_record.py:
```diff
@@ -0,0 +1,492 @@
+            r = dict(record)
+            del r['id']
+            # python3 does not have cmp so let's use the official workaround
+            if (r > self.payload) - (r < self.payload) == 0:
```
This blows up for me (on python 3) with:
```
/home/phemmer/.ansible/tmp/ansible-tmp-1547002269.7-214143530475396/AnsiballZ_digital_ocean_domain_record.py:17: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Traceback (most recent call last):
  File "/home/phemmer/.ansible/tmp/ansible-tmp-1547002269.7-214143530475396/AnsiballZ_digital_ocean_domain_record.py", line 113, in <module>
    _ansiballz_main()
  File "/home/phemmer/.ansible/tmp/ansible-tmp-1547002269.7-214143530475396/AnsiballZ_digital_ocean_domain_record.py", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/phemmer/.ansible/tmp/ansible-tmp-1547002269.7-214143530475396/AnsiballZ_digital_ocean_domain_record.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/usr/lib64/python3.7/imp.py", line 234, in load_module
    return load_source(name, filename, file)
  File "/usr/lib64/python3.7/imp.py", line 169, in load_source
    module = _exec(spec, sys.modules[name])
  File "<frozen importlib._bootstrap>", line 630, in _exec
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/tmp/ansible_digital_ocean_domain_record_payload_a66bc9an/__main__.py", line 492, in <module>
  File "/tmp/ansible_digital_ocean_domain_record_payload_a66bc9an/__main__.py", line 484, in main
  File "/tmp/ansible_digital_ocean_domain_record_payload_a66bc9an/__main__.py", line 334, in create_or_update_record
  File "/tmp/ansible_digital_ocean_domain_record_payload_a66bc9an/__main__.py", line 296, in __get_matching_records
TypeError: '>' not supported between instances of 'dict' and 'dict'
```

woohgit replied:
@phemmer This code hasn't been updated for 1.5 years. :( So the python 3.7 specific dict compare was not written / updated. In the meantime I had to realize that ansible is definitely not the right tool to manage domain records/entries so I'm using `doctl` and `terraform` to maintain the desired state instead. Although I used my own plugin to create tons of records 1.5 years ago.

phemmer replied:
I was able to fix this locally by changing to `if r == self.payload:`. No idea if there's some case this doesn't work for though.

At scale ansible might not be the right tool. However I have no desire to deal with adding another tool just for updating a single record.

woohgit commented:
We can call it dead. I wrote this module when I needed this feature but after I did the work I had to realize that Ansilbe is not good for such a task. Additionally, there was no feedback at all from any upstream devs so I let it die. I'm closing it out.
